### ISSUE TITLE
Name clash - @name v5.11

### ISSUE THREAD

**Issue by MSP-Greg** [seattlerb/minitest#729]
While working on testing RubyGems, had a name clash with `@name` while using 5.11.

21 Errors with one class, renamed the variable from `@name` to `@r_name`.

Might consider something similar here...

EDIT: One if my (sometimes forgotten) rules is to never use plain `name` for anything.  Always a prefix.  Mintest does seem to 'encourage' some things.  Regardless, I suspect there may be a few repos/gems broken due to the naming clash...

**Comment by zenspider** [comment]
While using 5.11? It's been this way since 5.0 iirc. Was there problems associated to a minor bump?

**Comment by MSP-Greg** [comment]
Previously, RubyGems, for various reasons (partly because Ruby does so), was using a 4.x version of Minitest.   

I thought it might be time to use a current version, and the tests became unstable.  Hence, I've been running the tests quite often.  So I start another Travis run, and I get [this](https://travis-ci.org/MSP-Greg/rubygems/builds/323957866).  Other than me having a typo, the tests have not ever failed on all jobs.  Also, between this build and the previous two (started via Travis 'trigger build'), there were no changes.

So, saw the class, saw the error count matched the `test_` method count, and saw the variable `@name`.  Gave it a prefix, ran the tests again, and got [this](https://travis-ci.org/MSP-Greg/rubygems/builds/323970876).  That was the **only** change.  There have been no other changes.

I haven't spent that much time thinking about it, but it would seem...

**Comment by zenspider** [comment]
Well `@name` is documented API for minitest 5 so it stands to reason that it'll bork. Borking semi-randomly seems odd, but who knows? Your prefix solution is a recommended solution. 

You also seem to have a double load error. You should make sure that your CI runs use `--disable-gems`.

Also, your green 2.5.0 run has a thread error printing in the middle of your output... so something is either noisy or an error is being swallowed and covered up, because the run itself is green.

**Comment by MSP-Greg** [comment]
@zenspider

> You also seem to have a double load error. You should make sure that your CI runs use `--disable-gems`.

Yes, it is using that, but I don't think it was available for ruby < 2.0, not sure.  95% of my work is using trunk.

> Also, your green 2.5.0 run has a thread error printing in the middle of your output... so something is either noisy or an error is being swallowed and covered up, because the run itself is green.

That was a rather last minute change/addition to 2.5 (and trunk).  See [ruby #14143](https://bugs.ruby-lang.org/issues/14143).  One quote (not by me) was:

"I am willing to help to reduce the extra output and failures in test-all,
but I would like a OK from Matz to try enabling Thread.report_on_exception by default."

IOW, it's appearing in a lot of tests...

**Comment by MSP-Greg** [comment]
@zenspider 

I did check this locally after I saw the Travis errors.  So, I did so again.  With 5.11 installed, I get the errors.  Removing 5.11, leaving 5.10.3, I don't have the errors.  I'm using 
```
ruby 2.6.0dev (2018-01-01 trunk 61537) [x64-mingw32]
```

When I get the RubyGems stuff finished, I'll investigate further...  Thanks, & Happy New Year, Greg

**Comment by zenspider** [comment]
> On Jan 1, 2018, at 20:58, MSP-Greg <notifications@github.com> wrote:
> 
> @zenspider
> 
> I did check this locally after I saw the Travis errors. So, I did so again. With 5.11 installed, I get the errors. Removing 5.11, leaving 5.10.3, I don't have the errors.

In my changelog:

> • ProgressReporter#prerecord now explicitly prints klass.name. Allows for fakers.

That’s the only thing I have in here that has to do with `name` that I can remember… but your failures look like this:

>   4) Error:
> TestGemSourceGit#a:
> NoMethodError: undefined method `a' for #<TestGemSourceGit:0x0000000002a6dc70>
>     /home/travis/.rvm/gems/ruby-2.5.0/gems/minitest-5.11.0/lib/minitest/test.rb:93:in `block (3 levels) in run'
>     /home/travis/.rvm/gems/ruby-2.5.0/gems/minitest-5.11.0/lib/minitest/test.rb:190:in `capture_exceptions'
>     /home/travis/.rvm/gems/ruby-2.5.0/gems/minitest-5.11.0/lib/minitest/test.rb:90:in `block (2 levels) in run’

which is from:

>     def run
>       with_info_handler do
>         time_it do
>           capture_exceptions do
>             before_setup; setup; after_setup
> 
>             self.send self.name
>           end
> 
>           TEARDOWN_METHODS.each do |hook|
>             capture_exceptions do
>               self.send hook
>             end
>           end
>         end
>       end
> 
>       Result.from self # per contract
>     end

which that send hasn’t changed in ages… 2013/04/24 to be exact. I’m more curious about why your stuff passes on 5.10.3 than why it fails on 5.11.0.

**Comment by y-yagi** [comment]
I encountered the same issue with Minitest 5.11.0.  If use `@name` instance variable, got an error. 

Reproduce script: 

```ruby   
# dummy_test.rb
require "minitest/autorun"

class DummyTest < Minitest::Test
  def setup
    @name = "title"
  end

  def test_dummy
    assert_equal "title", @name
  end
end
```
 

```bash 
$ bundle exec ruby dummy_test.rb
Run options: --seed 34057

# Running:

E

Finished in 0.000668s, 1496.4101 runs/s, 0.0000 assertions/s.

  1) Error:
DummyTest#title:
NoMethodError: undefined method `title' for #<DummyTest:0x000055b3dbc12048>
    /home/yaginuma/.rbenv/versions/2.5.0/lib/ruby/gems/2.5.0/gems/minitest-5.11.0/lib/minitest/test.rb:93:in `block (3 levels) in run'
    /home/yaginuma/.rbenv/versions/2.5.0/lib/ruby/gems/2.5.0/gems/minitest-5.11.0/lib/minitest/test.rb:190:in `capture_exceptions'
    /home/yaginuma/.rbenv/versions/2.5.0/lib/ruby/gems/2.5.0/gems/minitest-5.11.0/lib/minitest/test.rb:90:in `block (2 levels) in run'
    /home/yaginuma/.rbenv/versions/2.5.0/lib/ruby/gems/2.5.0/gems/minitest-5.11.0/lib/minitest.rb:265:in `time_it'
    /home/yaginuma/.rbenv/versions/2.5.0/lib/ruby/gems/2.5.0/gems/minitest-5.11.0/lib/minitest/test.rb:89:in `block in run'
    /home/yaginuma/.rbenv/versions/2.5.0/lib/ruby/gems/2.5.0/gems/minitest-5.11.0/lib/minitest.rb:360:in `on_signal'
    /home/yaginuma/.rbenv/versions/2.5.0/lib/ruby/gems/2.5.0/gems/minitest-5.11.0/lib/minitest/test.rb:206:in `with_info_handler'
    /home/yaginuma/.rbenv/versions/2.5.0/lib/ruby/gems/2.5.0/gems/minitest-5.11.0/lib/minitest/test.rb:88:in `run'
    /home/yaginuma/.rbenv/versions/2.5.0/lib/ruby/gems/2.5.0/gems/minitest-5.11.0/lib/minitest.rb:948:in `run_one_method'
    /home/yaginuma/.rbenv/versions/2.5.0/lib/ruby/gems/2.5.0/gems/minitest-5.11.0/lib/minitest.rb:334:in `run_one_method'
    /home/yaginuma/.rbenv/versions/2.5.0/lib/ruby/gems/2.5.0/gems/minitest-5.11.0/lib/minitest.rb:321:in `block (2 levels) in run'
    /home/yaginuma/.rbenv/versions/2.5.0/lib/ruby/gems/2.5.0/gems/minitest-5.11.0/lib/minitest.rb:320:in `each'
    /home/yaginuma/.rbenv/versions/2.5.0/lib/ruby/gems/2.5.0/gems/minitest-5.11.0/lib/minitest.rb:320:in `block in run'
    /home/yaginuma/.rbenv/versions/2.5.0/lib/ruby/gems/2.5.0/gems/minitest-5.11.0/lib/minitest.rb:360:in `on_signal'
    /home/yaginuma/.rbenv/versions/2.5.0/lib/ruby/gems/2.5.0/gems/minitest-5.11.0/lib/minitest.rb:347:in `with_info_handler'
    /home/yaginuma/.rbenv/versions/2.5.0/lib/ruby/gems/2.5.0/gems/minitest-5.11.0/lib/minitest.rb:319:in `run'
    /home/yaginuma/.rbenv/versions/2.5.0/lib/ruby/gems/2.5.0/gems/minitest-5.11.0/lib/minitest.rb:159:in `block in __run'
    /home/yaginuma/.rbenv/versions/2.5.0/lib/ruby/gems/2.5.0/gems/minitest-5.11.0/lib/minitest.rb:159:in `map'
    /home/yaginuma/.rbenv/versions/2.5.0/lib/ruby/gems/2.5.0/gems/minitest-5.11.0/lib/minitest.rb:159:in `__run'
    /home/yaginuma/.rbenv/versions/2.5.0/lib/ruby/gems/2.5.0/gems/minitest-5.11.0/lib/minitest.rb:136:in `run'
    /home/yaginuma/.rbenv/versions/2.5.0/lib/ruby/gems/2.5.0/gems/minitest-5.11.0/lib/minitest.rb:63:in `block in autorun'

1 runs, 0 assertions, 0 failures, 1 errors, 0 skips
```  

This works with Minitest 5.10.3. Thanks.

**Comment by MSP-Greg** [comment]
@zenspider

I won't be able to look at this until later, but this issue starts with commit 00433fc.  Commit 30eaa08 is okay.

I just symlinked from my ruby gem folder to the repo, did `git checkout`, and ran @y-yagi's example...

**Comment by zenspider** [comment]
OK. I'm clued in now. Thanks @y-yagi. That helped. The problem was that I shuffled around the class hierarchy and made Result a superclass of Test (it was originally sibling to it) and it added the `:name` accessors AGAIN... overriding those from `Runnable` (which were accessing `@NAME`). Doing the following is an acceptable patch until I get a bug fix out today:

```ruby
class Minitest::Result
  remove_method :name
  remove_method :name=
end if defined?(Minitest::Result) # make work on both 5.10 and 5.11
```

**Comment by MSP-Greg** [comment]
5e7d97b fixed the issue with RubyGems testing.  I'm okay to close...  Thanks.

**Comment by zenspider** [comment]
Released 5.11.1. This should be fixed. Thanks!
